actioncable/test/connection/stream_test.rb

```diff
@@ -0,0 +1,69 @@
+require 'test_helper'
+require 'stubs/test_server'
+
+class ActionCable::Connection::StreamTest < ActionCable::TestCase
+  class Connection < ActionCable::Connection::Base
+    attr_reader :websocket, :subscriptions, :message_buffer, :connected
+    attr_reader :errors
+
+    def initialize(*)
+      super
+      @errors = []
+    end
+
+    def connect
+      @connected = true
+    end
+
+    def disconnect
+      @connected = false
+    end
+
+    def send_async(method, *args)
+      send method, *args
+    end
+
+    def on_error(message)
+      @errors << message
+    end
+  end
+
+  setup do
+    @server = TestServer.new
+    @server.config.allowed_request_origins = %w( http://rubyonrails.com )
+  end
+
+  [[ EOFError, 'foo'],
+   [ Errno::ECONNRESET, 'Connection reset by peer - foo' ]
+  ].each do |closed_exception, message|
+    test "closes socket on #{closed_exception}" do
+      skip if ENV['FAYE'].present?
+
+      run_in_eventmachine do
+        connection = open_connection
+
+        # Internal hax = :(
+        client = connection.websocket.send(:websocket)
+        client.instance_variable_get('@stream').instance_variable_get('@rack_hijack_io').expects(:write).raises(closed_exception, 'foo')
+        client.expects(:client_gone)
+
+        client.write('boo')
+        assert_equal [message], connection.errors
+      end
+    end
+  end
+
+  private
+    def open_connection
+      env = Rack::MockRequest.env_for '/test',
+        'HTTP_CONNECTION' => 'upgrade', 'HTTP_UPGRADE' => 'websocket',
+        'HTTP_HOST' => 'localhost', 'HTTP_ORIGIN' => 'http://rubyonrails.com'
+      env['rack.hijack'] = -> { env['rack.hijack_io'] = StringIO.new }
+
+      Connection.new(@server, env).tap do |connection|
+        connection.process
+        connection.send :handle_open
+        assert connection.connected
+      end
+    end
+end
```

Comment:
Lots of duplication here. Figure we can eat it for now and incrementally get closer to 1. using the actual classes rather than test stubs and 2. disentangling the classes so they're easier to manipulate and use on their own.